src/hotspot/share/logging/logAsyncWriter.hpp

```diff
@@ -140,7 +55,79 @@ typedef ResourceHashtable<LogFileStreamOutput*,
 // change the logging configuration via jcmd, LogConfiguration::configure_output() calls flush() under the protection of the
 // ConfigurationLock. In addition flush() is called during JVM termination, via LogConfiguration::finalize.
 class AsyncLogWriter : public NonJavaThread {
+  friend class AsyncLogTest;
   class AsyncLogLocker;
+  using AsyncLogMap = ResourceHashtable<LogFileStreamOutput*,
+                          uint32_t,
+                          17, /*table_size*/
+                          ResourceObj::C_HEAP,
+                          mtLogging>;
+  class Message {
+    LogFileStreamOutput* const _output;
+    const LogDecorations _decorations;
+
+   public:
+    // only-valid for in-place new!
+    Message(LogFileStreamOutput* output, const LogDecorations& decorations, const char* msg)
+      : _output(output), _decorations(decorations) {
+      assert(msg != nullptr, "c-str message can not be null!");
+      PRAGMA_STRINGOP_OVERFLOW_IGNORED
+      strcpy(reinterpret_cast<char* >(this+1), msg);
+    }
+
+    size_t size() const {
+      size_t len = strlen(message()) + 1;
+      return align_up(sizeof(*this) + len, sizeof(void*));
+    }
+
+    bool is_token() const { return _output == nullptr; }
+    LogFileStreamOutput* output() const { return _output; }
+    const LogDecorations& decorations() const { return _decorations; }
+    const char* message() const { return reinterpret_cast<const char *>(this+1); }
+  };
+
+  // A flush TOKEN is Message(nullptr, None, "");
+  static const size_t TOKEN_SIZE = {align_up(sizeof(Message) + 1, sizeof(void*))};
+
+  class Buffer : public CHeapObj<mtLogging> {
+    size_t _pos;
+    char* _buf;
+    size_t _capacity;
+
+   public:
+    Buffer(size_t capacity);
+    ~Buffer();
+
+    void push_flush_token();
+    bool push_back(LogFileStreamOutput* output, const LogDecorations& decorations, const char* msg);
+
+    // for testing-only!
+    size_t set_capacity(size_t value) {
+      size_t old = _capacity;
+      _capacity = value;
+      return old;
+    }
+
+    void reset() { _pos = 0; }
+
+    class Iterator {
+      const Buffer& _buf;
+      size_t _curr;
+
+    public:
+      Iterator(const Buffer& buffer): _buf(buffer), _curr(0) {}
+
+      bool is_empty() const {
+        return _curr >= _buf._pos;
```

Comment:
This should never be larger than `_buf._pos`, right? Also, I would not name it `is_empty`, that's confusing, but rather something like `eof()` or `at_end()` or somesuch.